src/back/hlsl/writer.rs

```diff
@@ -72,42 +92,267 @@ impl<W: Write> Writer<W> {
+            };
+
+            let mut members = Vec::with_capacity(func.arguments.len());
+
+            write!(self.out, "struct {}", &struct_name)?;
+            writeln!(self.out, " {{")?;
+
+            for arg in func.arguments.iter() {
+                let arg_name = if let Some(ref name) = arg.name {
+                    name
+                } else {
+                    "member"
+                };
+                let member = EpStructMember {
+                    name: self.namer.call_unique(arg_name),
+                    ty: arg.ty,
+                    binding: arg.binding.clone(),
+                };
+
+                write!(self.out, "{}", INDENT)?;
+                self.write_type(module, member.ty)?;
+                write!(self.out, " {}", &member.name)?;
+                if let Some(ref binding) = member.binding {
+                    self.write_binding(binding)?;
+                }
+                write!(self.out, ";")?;
                 writeln!(self.out)?;
+
+                members.push(member);
             }
+
+            writeln!(self.out, "}};")?;
+            writeln!(self.out)?;
+
+            self.ep_input = Some(EntryPointBinding {
+                stage,
+                name: struct_name,
+                members,
+            });
+        } else {
+            log::error!("Entry point without input arguments in stage {:?}", stage);
         }
 
         Ok(())
     }
 
+    /// Helper method used to write global variables
+    /// # Notes
+    /// Always adds a newline
+    fn write_global(&mut self, module: &Module, handle: Handle<GlobalVariable>) -> BackendResult {
+        let global = &module.global_variables[handle];
+        let inner = &module.types[global.ty].inner;
+
+        let register_ty = match *inner {
+            TypeInner::Image { .. } => "t",
+            TypeInner::Sampler { .. } => "s",
+            _ => "",
+        };
+
+        let register = if let Some(ref binding) = global.binding {
+            format!("register({}{})", register_ty, binding.binding)
+        } else {
+            String::from("")
+        };
+
+        let name = self.names[&NameKey::GlobalVariable(handle)].clone();
+        self.write_type(module, global.ty)?;
+        writeln!(self.out, " {} : {};", name, register)?;
+
+        Ok(())
+    }
+
+    /// Helper method used to write global constants
+    ///
+    /// # Notes
+    /// Ends in a newline
+    fn write_global_constant(
+        &mut self,
+        _module: &Module,
+        inner: &ConstantInner,
+        handle: Handle<Constant>,
+    ) -> BackendResult {
+        match *inner {
+            ConstantInner::Scalar {
+                width: _,
+                ref value,
+            } => {
+                let name = &self.names[&NameKey::Constant(handle)];
+                let (ty, value) = match *value {
+                    crate::ScalarValue::Sint(value) => ("int", format!("{}", value)),
```

Comment:
It would be good to try to avoid heap allocations from the start.
On one hand, this is unlikely going to manifest too much in performance profiles, at least here specifically.
On the other hand, we have parties constantly looking to pre-compile HLSL and friends, which is a bit problematic. We don't need to give them more reason to do so (by making Naga slower).

Reply:
Would it be better if we make 2 times match? We can avoid heap allocation with something like this:

```rust
let ty = match *value {
    crate::ScalarValue::Sint(value) => "int",
...

match *value {
    crate::ScalarValue::Sint(value) => write!(..),
    crate::ScalarValue::Uint(value) => write!(..),
    crate::ScalarValue::Float(value) => write!(..),
```

Reply:
that sounds fine, yes